go/common/headers.go

```diff
@@ -64,17 +95,93 @@ type Withdrawal struct {
+	cp := *b
+	cp.R = nil
+	cp.S = nil
+	hash, err := rlpHash(cp)
+	if err != nil {
+		panic("err hashing batch header")
+	}
+	return hash
+}
+
+func (b *BatchHeader) ToRollupHeader() *RollupHeader {
+	return &RollupHeader{
+		ParentHash:                    b.ParentHash,
+		UncleHash:                     b.UncleHash,
+		Coinbase:                      b.Coinbase,
+		Root:                          b.Root,
+		TxHash:                        b.TxHash,
+		ReceiptHash:                   b.ReceiptHash,
+		Bloom:                         b.Bloom,
+		Difficulty:                    b.Difficulty,
+		Number:                        b.Number,
+		GasLimit:                      b.GasLimit,
+		GasUsed:                       b.GasUsed,
+		Time:                          b.Time,
+		Extra:                         b.Extra,
+		MixDigest:                     b.MixDigest,
+		Nonce:                         b.Nonce,
+		BaseFee:                       b.BaseFee,
+		Agg:                           b.Agg,
+		L1Proof:                       b.L1Proof,
+		R:                             b.R,
+		S:                             b.S,
+		Withdrawals:                   b.Withdrawals,
+		CrossChainMessages:            b.CrossChainMessages,
+		LatestInboudCrossChainHash:    b.LatestInboudCrossChainHash,
+		LatestInboundCrossChainHeight: b.LatestInboundCrossChainHeight,
+	}
+}
+
+// Hash returns the block hash of the header, which is simply the keccak256 hash of its
+// RLP encoding excluding the signature.
+func (r *RollupHeader) Hash() L2RootHash {
+	cp := *r
 	cp.R = nil
 	cp.S = nil
 	hash, err := rlpHash(cp)
 	if err != nil {
-		panic("err hashing a rollup header")
+		panic("err hashing rollup header")
 	}
 	return hash
 }
 
+func (r *RollupHeader) ToBatchHeader() *BatchHeader {
+	return &BatchHeader{
+		ParentHash:                    r.ParentHash,
+		UncleHash:                     r.UncleHash,
+		Coinbase:                      r.Coinbase,
+		Root:                          r.Root,
+		TxHash:                        r.TxHash,
+		ReceiptHash:                   r.ReceiptHash,
+		Bloom:                         r.Bloom,
+		Difficulty:                    r.Difficulty,
+		Number:                        r.Number,
+		GasLimit:                      r.GasLimit,
+		GasUsed:                       r.GasUsed,
+		Time:                          r.Time,
+		Extra:                         r.Extra,
+		MixDigest:                     r.MixDigest,
+		Nonce:                         r.Nonce,
+		BaseFee:                       r.BaseFee,
+		Agg:                           r.Agg,
+		L1Proof:                       r.L1Proof,
+		R:                             r.R,
+		S:                             r.S,
+		Withdrawals:                   r.Withdrawals,
+		CrossChainMessages:            r.CrossChainMessages,
+		LatestInboudCrossChainHash:    r.LatestInboudCrossChainHash,
+		LatestInboundCrossChainHeight: r.LatestInboundCrossChainHeight,
+	}
+}
+
+// Used to hash headers.
+var hasherPool = sync.Pool{
```

Comment:
minor (maybe subjective?) thing but I think global vars normally go to the top, above struct definitions